From 1597b37f9867bf73ba2f714f5820edaec11376ac Mon Sep 17 00:00:00 2001
From: Karthikeyan K <karthikk@habana.ai>
Date: Mon, 31 Jan 2022 05:16:12 +0200
Subject: [PATCH] [SW-68803] boot_desc: fix erom specific hardcoding

1) Fixing the hardcoded macro for max images in erom to refer from
image index enumeration. This is to improve forward compatibility.

2) hl_boot_desc_erom.h is renamed to hl_boot_desc_rom.h as the header
is no longer dedicated to erom. This header will now contain defines
specific for boot rom. Any common definitions that should be shared
with ppboot/preboot will refer the specs header for boot descriptors.

3) Referring the image descriptor type from common hl_boot_desc.h
specs file.

Change-Id: I2c5cde51dae2fa2eb9d8603a725323701334ae95
---

diff --git a/app/arcboot_rom.c b/app/arcboot_rom.c
index ecd2e36..58b9e20 100644
--- a/app/arcboot_rom.c
+++ b/app/arcboot_rom.c
@@ -21,11 +21,11 @@
 	.validation_pass = 0,
 	.efuse_read = 0,
 	.curr_img = 0,
-	.erom_boot_desc_flag = 0,
+	.rom_boot_desc_flag = 0,
 	.die_id = 0,
 	.sku = 0,
 	.fuse_spi_offset = 0x0,
-	.fuse_spi = false
+	.fuse_spi = false,
 };
 
 #define ARCBOOT_BRINGUP_ERROR		-1
diff --git a/arch/gaudi2/arc/common.c b/arch/gaudi2/arc/common.c
index dad125a..255d156 100644
--- a/arch/gaudi2/arc/common.c
+++ b/arch/gaudi2/arc/common.c
@@ -288,7 +288,7 @@
 		return;
 	}
 
-	if (!arcboot_rom_global.erom_boot_desc_flag)
+	if (!arcboot_rom_global.rom_boot_desc_flag)
 		if (!boot_desc_init()) {
 			arcboot_rom_status_update(
 					PROGRESS_ARCBOOT_EROM_BOOT_DESC_ERROR);
@@ -314,7 +314,7 @@
 	/* check fuse bit to call the erom function*/
 	if ((arcboot_rom_global.config_fuse &
 	   EFUSE_SYSCFG_ENABLE_TPM_MASK) != 0) {
-		if (!arcboot_rom_global.erom_boot_desc_flag)
+		if (!arcboot_rom_global.rom_boot_desc_flag)
 			if (!boot_desc_init()) {
 				arcboot_rom_status_update(
 					PROGRESS_ARCBOOT_EROM_BOOT_DESC_ERROR);
diff --git a/common/boot_desc.c b/common/boot_desc.c
index d3fd55e..a97cfee 100644
--- a/common/boot_desc.c
+++ b/common/boot_desc.c
@@ -8,7 +8,7 @@
 #include <string.h>
 #include <stdbool.h>
 #include <hl_fw_if.h>
-#include "hl_boot_desc_erom.h"
+#include "hl_boot_desc_rom.h"
 #include <common.h>
 #include <arcboot.h>
 #include "boot_desc.h"
@@ -22,9 +22,9 @@
 #define INV_VER1		((uint32_t)0)
 #define INV_VER2		((uint32_t) -1)
 #define INVALID_DESC_ID		2
-#define MAX_EROM_BINS		3
+
 static struct hl_fw_boot_ptrs *p_boot_ptrs;
-static struct hl_fw_img_desc *p_img_desc[MAX_EROM_BINS];
+static struct hl_fw_img_desc *p_img_desc[HL_FW_MAX_EROM_BOOT_PTRS - 1];
 
 static bool boot_desc_img_desc_is_valid(struct hl_fw_img_desc *img_desc)
 {
@@ -48,8 +48,7 @@
 
 	boot_ptrs_count = spi_flash_readl(HL_BOOT_PTS_COUNT_FLASH_OFFSET);
 
-	/* Expect just 'MAX_EROM_BINS' pointers, else give up */
-	if (boot_ptrs_count > MAX_EROM_BINS)
+	if (boot_ptrs_count >= HL_FW_MAX_EROM_BOOT_PTRS)
 		return 0;
 
 	img_ptrs_size = BOOT_PTRS_SZ + ((sizeof(uint32_t) +
@@ -79,7 +78,7 @@
 		if (!boot_desc_img_desc_is_valid(p_img_desc[i]))
 			return 0;
 	}
-	arcboot_rom_global.erom_boot_desc_flag = 1;
+	arcboot_rom_global.rom_boot_desc_flag = 1;
 	return p_boot_ptrs->count;
 }
 
@@ -93,7 +92,7 @@
 		return HL_FW_IMG_NOT_FOUND;
 
 	for (i = 0; i < p_boot_ptrs->count; i++) {
-		if (p_img_desc[i]->magic.magic_type.type != type)
+		if (p_img_desc[i]->magic.type != type)
 			continue;
 
 		desc->spi_offset = p_img_desc[i]->offset[p_img_desc[i]->idx];
diff --git a/common/boot_desc.h b/common/boot_desc.h
index dd90f19..d62e1c9 100644
--- a/common/boot_desc.h
+++ b/common/boot_desc.h
@@ -8,12 +8,12 @@
 #define _BOOT_DESC_H_
 
 #include <hl_fw_if.h>
-#include <hl_boot_desc_erom.h>
+#include <hl_boot_desc_rom.h>
 #include <sha512.h>
 
 #define HL_FW_IMG_NOT_FOUND		((uint32_t) -1)
 /* erom image pointer binary placed in flash at offset 0 */
-#define HL_EROM_IMG_PTRS_BASE		(0x00000000)
+#define HL_EROM_IMG_PTRS_BASE		(0x0)
 #define HL_CRC32_FOOTER			(sizeof(uint32_t))
 /* erom boot descriptors placed in SP SRAM at offset 4K */
 #define HL_BOOT_DESC_SP_SRAM_OFFSET	(4 * 1024)
diff --git a/common/hl_boot_desc_erom.h b/common/hl_boot_desc_erom.h
deleted file mode 100644
index 0a4b817..0000000
--- a/common/hl_boot_desc_erom.h
+++ /dev/null
@@ -1,103 +0,0 @@
-/* SPDX-License-Identifier: MIT
- *
- * Copyright 2021 HabanaLabs, Ltd.
- * All Rights Reserved.
- *
- */
-
-#ifndef HL_BOOT_DESC_EROM_H
-#define HL_BOOT_DESC_EROM_H
-
-#include <stdint.h>
-
-/*
- * !!!! IMPORTANT INFO !!!!
- * EROM BOOT DESCRIPTORS (448 BYTES) ARE LOADED INTO SP SRAM at offset 4K
- * HENCE THIS ADDRESS SHOULD NOT BE USED FOR OTHER PURPOSES.
- * 448 = sizeof(hl_fw_boot_ptrs) +
- *       (3 * sizeof(uint32_t) +
- *       (3 * sizeof(hl_fw_img_desc))
- * where 3 is the number of erom binaries.
- */
-
-/* FW boot pointers magic:	  N X T 0 */
-#define HL_FW_BOOT_MAGIC	0x4E585400
-
-/* FW image descriptor magic:	  P T R 0 | <hl_fw_img_type> */
-#define HL_FW_IMG_DESC_MAGIC	0x50545200
-
-/* Current version of the image descriptor */
-#define HL_FW_IMG_DESC_CUR_VER	1
-
-#define HL_CHKSUM_SIZE		48
-
-enum hl_fw_img_type {
-	HL_FW_IMG_ROOT = 0,
-	HL_FW_EROM_PCIE = 1,
-	HL_FW_EROM_RECOVERY = 2,
-	HL_FW_EROM_TPM = 3,
-};
-
-struct boot_magic_type {
-	uint8_t type;	/* enum hl_fw_img_type */
-	char name[3];
-} __packed;
-
-struct hl_fw_boot_magic {
-	union {
-		struct boot_magic_type magic_type;
-		uint32_t val;
-	};
-};
-
-/*
- *      struct hl_fw_boot_ptrs     struct hl_fw_img_desc
- *           ============            =================
- *          || Next     ||    +---->|| Magic         ||
- *          | ---------- |    |     | --------------- |
- *          || Mod-able ||    |     || Version       ||
- *          | ---------- |    |     | --------------- |
- *          || Magic    ||    |     || Index         ||
- *          | ---------- |    |     | --------------- |
- *     +--<<|| Count    ||    |     || Offset    [2] ||
- *     |     ============     |     | --------------- |
- *     |                      |     || Size      [2] ||
- *     |                      |     | --------------- |
- *     |                      |     || Load High [2] ||
- *     |                      |     | --------------- |
- *     |                      |     || Load Low  [2] ||
- *     |                      |     | --------------- |
- *     |                      |     || SHA256    [2] ||
- *     |    Following Array   |     | --------------- |
- *     |     ============     |      =================
- *     +--->|| Offset   ||>>--+
- *     |    | ---------- |           =================
- *     +--->|| Offset   ||>>------->|  ...            |
- *     |    | ---------- |           =================
- *     :       :            :               :
- *     |    | ---------- |           =================
- *     +--->|| Offset   ||>>------->|  ...            |
- *           ============            =================
- *
- */
-
-struct hl_fw_boot_ptrs {
-	uint32_t next;	/* backward compatibility */
-	uint32_t mod;	/* backward compatibility */
-	struct hl_fw_boot_magic magic;
-	uint32_t count;
-	/* array of offsets by count size must follow w/o gaps */
-};
-
-struct hl_fw_img_desc {
-	struct hl_fw_boot_magic magic;
-	uint32_t desc_ver;	/* descriptor version */
-	uint32_t idx;		/* "next" index inside the offset array */
-	uint32_t offset[2];	/* the offset on the SPI flash */
-	uint32_t size[2];	/* the size of the image to load */
-	uint32_t load_high[2];	/* the high [32-63] bits of the load address */
-	uint32_t load_low[2];	/* the low [0-31] bits of the load address */
-	uint8_t chksum[2][HL_CHKSUM_SIZE];	/* checksum for image */
-};
-
-#endif /* HL_BOOT_DESC_EROM_H */
diff --git a/common/hl_boot_desc_rom.h b/common/hl_boot_desc_rom.h
new file mode 100644
index 0000000..4bd1a73
--- /dev/null
+++ b/common/hl_boot_desc_rom.h
@@ -0,0 +1,22 @@
+/* SPDX-License-Identifier: MIT
+ *
+ * Copyright 2021 HabanaLabs, Ltd.
+ * All Rights Reserved.
+ *
+ */
+
+#ifndef HL_BOOT_DESC_ROM_H
+#define HL_BOOT_DESC_ROM_H
+
+#include <stdint.h>
+#include <hl_boot_desc.h>
+
+enum hl_fw_img_type_erom {
+	HL_FW_EROM_IMG_ROOT = 0,
+	HL_FW_EROM_PCIE = 1,
+	HL_FW_EROM_RECOVERY = 2,
+	HL_FW_EROM_TPM = 3,
+	HL_FW_MAX_EROM_BOOT_PTRS,
+};
+
+#endif /* HL_BOOT_DESC_ROM_H */
diff --git a/include/arcboot.h b/include/arcboot.h
index a7584be..5c27c79 100644
--- a/include/arcboot.h
+++ b/include/arcboot.h
@@ -65,7 +65,7 @@
 	uint8_t validation_pass;
 	uint8_t efuse_read;
 	uint8_t curr_img;
-	uint8_t erom_boot_desc_flag;
+	uint8_t rom_boot_desc_flag;
 	uint8_t die_id;
 	uint8_t sku; /* 0: Dual die 1: Single die */
 	uint32_t fuse_spi_offset;
